src/Covenant/Internal/Expr.hs

```diff
@@ -0,0 +1,268 @@
+      aLam :: Int -> GenT (Reader Word64) (ExprBuilder Id)
+      aLam size = do
+        body <- expand (argOrId size)
+        -- Note (Koz, 21/01/25): We 'bypass' our protections on construction of
+        -- lambdas here for convenience, as otherwise, we would have to carry
+        -- around a `Scope n`, with `n` changing. While not impossible, this
+        -- would require existentialization (using something like `Some`), which
+        -- doesn't really help us any (because we have full control anyway), but
+        -- does make the code much harder to read and debug.
+        pure $ body >>= \bodyRef -> idOf (Lam bodyRef)
+      -- Note (Koz, 21/01/25): This is essentially `local`, but because GenT
+      -- isn't an instance of MonadReader, we have to write it by hand. We could
+      -- have defined an orphan instance instead, but I figured it'd be better
+      -- not to.
+      expand ::
+        forall (a :: Type).
+        GenT (Reader Word64) (ExprBuilder a) -> GenT (Reader Word64) (ExprBuilder a)
+      expand (GenT f) = GenT $ \qcgen size -> local (+ 1) (f qcgen size)
+      anApp :: Int -> GenT (Reader Word64) (ExprBuilder Id)
+      anApp size = do
+        lhs <- aLam size
+        rhs <- argOrId size
+        -- Note (Koz, 21/01/25): This is technically not as general as it could
+        -- be. By 'recycling' aLam, we always generate a 'fresh' lambda, rather
+        -- than (possibly) re-using one available to us as an argument. This is
+        -- not really practical to track at the moment, though once we have
+        -- types, it could be done. For now, this works well enough.
+        pure $ lhs >>= \lhsRef -> rhs >>= \rhsRef -> app (AnId lhsRef) rhsRef
+      argOrId :: Int -> GenT (Reader Word64) (ExprBuilder Ref)
+      argOrId originalSize = do
+        availableArgs <- lift ask
+        if availableArgs == 0
+          -- No arguments possible, so reduce the size and recurse
+          then fmap AnId <$> go (originalSize `quot` 2)
+          -- 20% chance we make an argument, 80% chance we don't
+          else
+            frequency
+              [ (1, pure . AnArg . Arg <$> liftGen (chooseEnum (0, availableArgs - 1))),
+                (4, fmap AnId <$> go (originalSize `quot` 2))
+              ]
+
+-- | Construct a literal (constant) value.
+--
+-- @since 1.0.0
+lit :: AConstant -> ExprBuilder Id
+lit = idOf . Lit
+
+-- | Construct a primitive function call.
+--
+-- @since 1.0.0
+prim :: PrimCall -> ExprBuilder Id
+prim = idOf . Prim
+
+-- | Construct a function application. The first argument is (an expression
+-- evaluating to) a function, the second argument is (an expression evaluating
+-- to) an argument.
+--
+-- = Important note
+--
+-- Currently, this does not verify that the first argument is indeed a function,
+-- nor that the second argument is appropriate.
+app :: Ref -> Ref -> ExprBuilder Id
+app f x = idOf (App f x)
+
+idOf :: Expr -> ExprBuilder Id
```

Comment:
Am I understanding this correctly. ExprBuilder is some sort of a symbol table where we enforce that each symbol/expression is defined only once?

And this `idOf` is adding an expression to the symbol table if missing.

In this case `idOf` doesn't seem to me as an appropriate name - I would maybe prefer something like `getOrAddIfMissing` but that is a code smell IMO. 
A cleaner solution would be to have a function purely for getting and purely for inserting..

If you don't agree, just add a proper description please because it's a bit confusing.

Reply:
The `Bimap` state ensures that hash consing works. Separating lookup and insertion is possible certainly, but the whole idea is that you can never 're-use' an `Id`. Separate operations would permit such re-use, which is undesirable.

I will think of a better name and write some explanation.